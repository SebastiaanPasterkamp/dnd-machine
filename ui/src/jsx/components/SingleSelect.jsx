import React from 'react';
import PropTypes from 'prop-types';
import LazyComponent from '../components/LazyComponent.jsx';

import _ from 'lodash';

import BaseSelect from './BaseSelect.jsx';

class SingleSelect extends LazyComponent
{
    getId(item) {
        if ('code' in item) {
            return item.code;
        }
        return item.name;
    }

    getName(item) {
        if ('label' in item) {
            return item.label;
        }
        return item.name;
    }

    onClick(item) {
        if (this.props.isDisabled(item)) {
            return;
        }

        this.props.setState(this.getId(item));
    }

    getLabel() {
        let label = this.props.emptyLabel;
        let item = _.find(
            this.props.items,
            {code: this.props.selected}
        ) || _.find(
            this.props.items,
            {name: this.props.selected}
        );
        if (
            !_.isNil(item)
            && !_.isNil(this.getId(item))
        ) {
            label = this.getName(item);
        }
        return label;
    }

    renderItem(item) {
        if (_.isNil(this.getId(item))) {
            return null;
        }
        let isDisabled = this.props.isDisabled(item);
        let style = _.filter([
            this.getId(item) == this.props.selected
                ? "info"
                : null,
            isDisabled
                ? "disabled"
                : null
            ]);
        return <li
                key={this.getId(item)}
                className={style.length ? style.join(' ') : null}
                data-value={this.getId(item)}
                onClick={isDisabled
                    ? null
                    : () => this.onClick(item)
                }
                >
            <a>{this.getName(item)}</a>
        </li>
    }

    render() {
        return <BaseSelect
                label={this.getLabel()}
                {...this.props}>
            {_.map(
                this.props.items,
                item => this.renderItem(item)
            )}
        </BaseSelect>;
    }
}

SingleSelect.defaultProps = {
    isDisabled: (item) => {
        return (
            'disabled' in item
            && item.disabled
        );
    },
    setState: (selected) => {
        console.log(['SingleSelect', selected]);
    }
};

SingleSelect.propTypes = {
    isDisabled: PropTypes.func.isRequired,
    setState: PropTypes.func.isRequired,
    emptyLabel: PropTypes.string,
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    selected: PropTypes.any,

};

export default SingleSelect;