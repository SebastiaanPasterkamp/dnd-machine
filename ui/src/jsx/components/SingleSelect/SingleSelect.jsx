import React from 'react';
import PropTypes from 'prop-types';
import {
    find,
    isNil,
    map,
} from 'lodash/fp';

import BaseSelect from '../BaseSelect';
import utils, { memoize } from '../../utils';

import SelectItem from './components/SelectItem';

class SingleSelect extends React.Component
{
    constructor(props) {
        super(props);
        this.state = {};
        this.memoize = memoize.bind(this);
    }

    static getDerivedStateFromProps(props, state) {
        const { defaultValue, selected, setState } = props;
        if (isNil(selected) && !isNil(defaultValue)) {
            setState(defaultValue);
        }
        return null;
    }

    onClick(item) {
        const { id } = item || {};
        return this.memoize(id, () => {
            const { objects, setState } = this.props;
            if (objects) {
                setState(item);
            } else {
                setState(id);
            }
        });
    }

    getLabel() {
        const {
            items,
            selected,
            objects,
            renderEmpty,
            emptyLabel,
        } = this.props;

        if (renderEmpty && selected === null) {
            return renderEmpty;
        }

        const { name } = objects
            ? (selected || {})
            : (
                find({ id: selected }, items)
                || find({ name: selected }, items)
                || { name: emptyLabel }
            );

        return name;
    }

    render() {
        const {
            items,
            selected,
            objects,
            setState,
            isDisabled,
            emptyLabel,
            renderEmpty,
            defaultValue,
            ...props,
        } = this.props;

        const { name: current = null } = objects
            ? (selected || {})
            : { name : selected };

        return (
            <BaseSelect
                label={ this.getLabel() }
                {...props}
            >
                {renderEmpty ? (
                    <SelectItem
                        id={null}
                        name={renderEmpty}
                        selected={current === null}
                        onClick={this.onClick(null)}
                    />
                ) : null}
                {map(item => {
                    const { id, name } = item;
                    return (
                        <SelectItem
                            key={id}
                            id={id}
                            name={name}
                            selected={current === id}
                            disabled={isDisabled(item)}
                            onClick={this.onClick(item)}
                        />
                    );
                })(items)}
            </BaseSelect>
        );
    }
}

SingleSelect.propTypes = {
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    selected: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.bool,
        PropTypes.shape({
            id: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string,
            ]).isRequired,
            name: PropTypes.string.isRequired,
        }),
    ]),
    objects: PropTypes.bool,
    setState: PropTypes.func.isRequired,
    isDisabled: PropTypes.func.isRequired,
    emptyLabel: PropTypes.string,
    renderEmpty: PropTypes.string,
};

SingleSelect.defaultProps = {
    isDisabled: (item) => item.disabled,
    selected: null,
    objects: false,
    defaultValue: null,
    renderEmpty: null,
    setState: (selected) => {
        console.log(['SingleSelect', selected]);
    },
};

export default SingleSelect;
