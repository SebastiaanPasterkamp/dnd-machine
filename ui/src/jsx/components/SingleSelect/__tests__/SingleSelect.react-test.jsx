import React from 'react';
import { mount } from 'enzyme';
import 'jest-enzyme';
import renderer from 'react-test-renderer';
import { stub } from 'sinon';

import SingleSelect from '../SingleSelect';

const props = {
    items: [
        {id: 1, name: 'One'},
        {id: 2, name: 'Two'},
        {id: 3, name: 'Three'},
    ],
};

const disabled = {
    items: [
        {id: 1, name: 'One', disabled: true},
        {id: 2, name: 'Two'},
        {id: 3, name: 'Three'},
    ],
};

describe('Component: SingleSelect', () => {
    it('should show a simple dropdown with 3 items', () => {
        const tree = renderer.create(
            <SingleSelect
                {...props}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should show a dropdown with 3 items in various states', () => {
        const tree = renderer.create(
            <SingleSelect
                {...disabled}
                selected={3}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should show a dropdown with 3 items disabled by callback', () => {
        const tree = renderer.create(
            <SingleSelect
                {...props}
                selected={3}
                isDisabled={(item) => { return item.id != 2; }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle unknown value', () => {
        const tree = renderer.create(
            <SingleSelect
                {...props}
                selected={4}
            />
        );

        expect(tree)
            .toMatchSnapshot();
    });

    it('should contain expected elements', () => {
        const wrapper = mount(
            <SingleSelect
                {...disabled}
                selected={2}
            />
        );

        expect(
            wrapper.find('button.nice-btn').exists()
        ).toBe(true);
        expect(
            wrapper.find('button.nice-btn').text()
        ).toEqual('Two');
        expect(
            wrapper.find('ul.dropdown-menu').exists()
        ).toBe(true);
        expect(
            wrapper.find('li.info').exists()
        ).toBe(true);
        expect(
            wrapper.find('li.disabled').exists()
        ).toBe(true);
    });

    it('should close on click', () => {
        const wrapper = mount(
            <SingleSelect
                {...disabled}
            />
        );

        expect(wrapper.find('ul.shown').exists()).toBe(false);

        wrapper.find('button').simulate('click');
        expect(wrapper.find('ul.shown').exists()).toBe(true);

        wrapper.find({'data-value': 1}).simulate('click');
        wrapper.update();
        expect(wrapper.find('ul.shown').exists()).toBe(false);
    });

    it('should callback for disabled states', () => {
        const onIsDisabled = stub();
        const wrapper = mount(
            <SingleSelect
                {...disabled}
                isDisabled={onIsDisabled}
            />
        );

        expect(onIsDisabled.callCount).toEqual(3);
    });

    it('should callback the newly selected item', () => {
        const onClickCallback = stub();
        const wrapper = mount(
            <SingleSelect
                {...disabled}
                setState={onClickCallback}
            />
        );

        wrapper.find('button').simulate('click');
        wrapper.find({'data-value': 2}).simulate('click');
        expect(onClickCallback.callCount).toEqual(1);
        expect(onClickCallback.calledWith(2)).toBe(true);
    });

    it('should not callback for disabled item', () => {
        const onClickCallback = stub();
        const wrapper = mount(
            <SingleSelect
                {...disabled}
                setState={onClickCallback}
            />
        );

        wrapper.find('button').simulate('click');
        wrapper.find({'data-value': 1}).simulate('click');
        expect(onClickCallback.callCount).toEqual(0);
    });
});
