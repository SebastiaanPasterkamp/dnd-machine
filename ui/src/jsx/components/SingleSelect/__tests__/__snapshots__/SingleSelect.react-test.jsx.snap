// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: SingleSelect should handle unknown value 1`] = `
<div
  className="nice-dropdown nice-form-control"
>
  <button
    className="nice-btn"
    disabled={false}
    onClick={[Function]}
  >
    
    <i
      className="icon fa-angle-down"
    />
  </button>
  <ul
    className="dropdown-menu"
    onClick={[Function]}
  >
    <li
      className={null}
      data-value={1}
      onClick={[Function]}
    >
      <a>
        One
      </a>
    </li>
    <li
      className={null}
      data-value={2}
      onClick={[Function]}
    >
      <a>
        Two
      </a>
    </li>
    <li
      className={null}
      data-value={3}
      onClick={[Function]}
    >
      <a>
        Three
      </a>
    </li>
  </ul>
</div>
`;

exports[`Component: SingleSelect should show a dropdown with 3 items disabled by callback 1`] = `
<div
  className="nice-dropdown nice-form-control"
>
  <button
    className="nice-btn"
    disabled={false}
    onClick={[Function]}
  >
    Three
    <i
      className="icon fa-angle-down"
    />
  </button>
  <ul
    className="dropdown-menu"
    onClick={[Function]}
  >
    <li
      className="disabled"
      data-value={1}
      onClick={null}
    >
      <a>
        One
      </a>
    </li>
    <li
      className={null}
      data-value={2}
      onClick={[Function]}
    >
      <a>
        Two
      </a>
    </li>
    <li
      className="info disabled"
      data-value={3}
      onClick={null}
    >
      <a>
        Three
      </a>
    </li>
  </ul>
</div>
`;

exports[`Component: SingleSelect should show a dropdown with 3 items in various states 1`] = `
<div
  className="nice-dropdown nice-form-control"
>
  <button
    className="nice-btn"
    disabled={false}
    onClick={[Function]}
  >
    Three
    <i
      className="icon fa-angle-down"
    />
  </button>
  <ul
    className="dropdown-menu"
    onClick={[Function]}
  >
    <li
      className="disabled"
      data-value={1}
      onClick={null}
    >
      <a>
        One
      </a>
    </li>
    <li
      className={null}
      data-value={2}
      onClick={[Function]}
    >
      <a>
        Two
      </a>
    </li>
    <li
      className="info"
      data-value={3}
      onClick={[Function]}
    >
      <a>
        Three
      </a>
    </li>
  </ul>
</div>
`;

exports[`Component: SingleSelect should show a simple dropdown with 3 items 1`] = `
<div
  className="nice-dropdown nice-form-control"
>
  <button
    className="nice-btn"
    disabled={false}
    onClick={[Function]}
  >
    
    <i
      className="icon fa-angle-down"
    />
  </button>
  <ul
    className="dropdown-menu"
    onClick={[Function]}
  >
    <li
      className={null}
      data-value={1}
      onClick={[Function]}
    >
      <a>
        One
      </a>
    </li>
    <li
      className={null}
      data-value={2}
      onClick={[Function]}
    >
      <a>
        Two
      </a>
    </li>
    <li
      className={null}
      data-value={3}
      onClick={[Function]}
    >
      <a>
        Three
      </a>
    </li>
  </ul>
</div>
`;
